proposals/0000-type-indexed-typeable.rst

```diff
@@ -0,0 +1,473 @@
+
+    -- | Information about a type constructor. No means of constructing 'TyCon's
+    -- is provided; the only values of this type available are those from
+    -- 'TypeRep's.
+    data TyCon
+    tyConPackage :: TyCon -> String
+    tyConModule :: TyCon -> String
+    tyConName :: TyCon -> String
+
+Type application can also be decomposed,
+
+.. code-block:: haskell
+
+    -- | A representation of a type application, @a b@. This is a bidirectional pattern.
+    pattern TRApp :: forall k2 (fun :: k2). ()
+                  => forall k1 (a :: k1 -> k2) (b :: k1). (fun ~ a b)
+                  => TypeRep a -> TypeRep b -> TypeRep fun
+
+We can also decompose function types (e.g. ``Int -> String``) in their argument
+(e.g. ``Int``) and result types (``String``). Strictly speaking this can be
+expressed in terms of ``TRApp`` but it seems like a common enough pattern that
+it's worth providing a pattern for it,
+
+.. code-block:: haskell
+
+    pattern TRFun :: forall fun. ()
+                  => forall arg res. (fun ~ (arg -> res))
+                  => TypeRep arg
+                  -> TypeRep res
+                  -> TypeRep fun
+
+We can also test for type equality,
+
+.. code-block:: haskell
+
+    -- | Kind-homogenous type equality
+    eqTypeRep  :: forall k (a :: k) (b :: k).
+                  TypeRep a -> TypeRep b -> Maybe (a :~: b)
+
+    -- | Kind-heterogenous type equality
+    eqTypeRep' :: forall k1 k2 (a :: k1) (b :: k2).
+                  TypeRep a -> TypeRep b -> Maybe (a :~~: b)
+
+Since ``TypeRep`` is a singleton, we can provide a means of satisfying a
+``Typeable`` constraint with a ``TypeRep`` without loss of coherence,
+
+.. code-block:: haskell
+
+    withTypeable :: TypeRep a -> (Typeable a => b) -> b
+
+We will also see later that it is helpful to quantify over the type index. For
+this we introduce,
+
+
+.. code-block:: haskell
+
+    data SomeTypeRep where
+        SomeTypeRep :: forall a. TypeRep a -> SomeTypeRep
+
+    instance Eq SomeTypeRep
+    instance Ord SomeTypeRep
+    instance Show SomeTypeRep
+
+    someTypeRep :: Typeable a => Proxy a -> SomeTypeRep
+
+Implementing ``Data.Dynamic``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+With this reflection machinery we can implement the ``Data.Dynamic`` type
+described in the Motivation section in a perfectly type-safe manner,
+
+.. code-block:: haskell
+
+    data Dynamic where
+        Dynamic :: TypeRep a -> a -> Dynamic
+
+    toDyn :: Typeable a => a -> Dynamic
+    toDyn v = Dynamic (typeOf v) v
+
+    fromDynamic :: Typeable a => Dynamic -> Maybe a
+    fromDynamic (Dynamic t v) =
+      case v of 
+        r | t `eqTypeRep` typeOf r -> Just r
+          | otherwise     -> Nothing
+       where
+
+    dynApply :: Dynamic -> Dynamic -> Maybe Dynamic
+    dynApply (Dynamic t1 f) (Dynamic t2 x) =
+      case funResultTy t1 t2 of
+        Just t3 -> Just (Dynamic t3 ((unsafeCoerce f) x))
+        Nothing -> Nothing
```

Comment:
This implementation of `dynApply` doesn't seem right, especially since it uses `unsafeCoerce`. Perhaps this definition was copy-pasted from above and wasn't updated? For reference, the definition in [your branch](https://github.com/bgamari/ghc/blob/594e9480fcf07b7fe6bf4c3c51918e8797b79ade/libraries/base/Data/Dynamic.hs#L134-L137) is:

```haskell
dynApply :: Dynamic -> Dynamic -> Maybe Dynamic
dynApply (Dynamic (TRFun ta tr) f) (Dynamic ta' x)
  | Just HRefl <- ta `eqTypeRep` ta'              = Just (Dynamic tr (f x))
dynApply _ _ = Nothing
```